### ISSUE TITLE
Add and update existing objects in collection

### ISSUE THREAD

**Issue by lkuendig** [mapstruct/mapstruct#3742]
### Use case

We want to update existing objects (entities) in a collection instead of using the "clear & addAll"-approach. Similar to https://stackoverflow.com/a/47296282. Therefore a distinction would be needed, which objects we want to update and which we want to add/remove.

Nice would be, to have a flag, which controls either to add or update. A "equals/contains" comparison would be sufficient.

### Generated Code

Flag = true on @Mapping

```
method (source, target) {
  cSource = copy(source)
  loop (target) {
    if (source.contains(t)) {
      update(source[t], t)
      cSource.remove(t)
    } else {
      target.remove(t)
    }
  }
  target.addAll(cSource)
}
```

in case this flag is enabled and there is no update method -> throw

### Possible workarounds

Currently we have implemented this approach by ignoring the collection-property and then implement the custom logic with @BeforeMapping.

This is quite a lot of code each time (just for a equals-comparison), error prone and not understood by someone who doesn't know the code (and therefore doesn't check for ignore & @BeforeMapping combination)

### MapStruct Version

1.6.0

**Comment by filiphr** [comment]
This is not super trivial to achieve with MapStruct. Why should we use `equals` to determine that the object matches, maybe there is another approach that should be used for this. I don't know much about your examples.

If you share a bit better example with how you are doing this with some actual types (that might be nested) I might be able to help out more. What I have in mind in the moment is that you could do something like:

```java
public interface UpdateMapper<T> {

    void update(T source, @MappingTarget T target);

    default void update(Collection<T> source, @MappingTarget Collection<T> target) {
        target.removeIf(t -> !source.contains(t));
        for (T sourceItem : source) {
            for (T targetItem : target) {
                if (targetItem.equals(sourceItem)) {
                    update(sourceItem, targetItem);
                }
            }
        }
    }
}
```

and then for your entities you will have those mappers and would be used.

Something that might be more interesting for you might be #1637. With that you might be able to improve is to then have a single static method e.g.

```java
static <V> void updateValues(Collection<V> source, @MappingTarget Collection<V> target, @MapStructTBD BiConsumer<V, V> updateFunction) {
        target.removeIf(t -> !source.contains(t));
        for (V sourceItem : source) {
            for (V targetItem : target) {
                if (targetItem.equals(sourceItem)) {
                    updateFunction.accept(sourceItem, targetItem);
                }
            }
        }
    }
```

MapStruct would then be able to call this method for all your collections and pass a single update method to that `BiConsumer`. That `BiConsumer` could be a `BiFunction`

**Comment by lkuendig** [comment]
In our case we use it for entities and in all cases it's sufficient to check for equality by the equals-method. But yes i can imagine there are cases where a custom "equals"-method would be great (a `equals`-function flag to reference a method?).
In general I would say, a entity should check by equals-method, value objects should not be updated, and for DTO's the equals method is sufficient in most cases but might require a custom method in special cases

Currently we have this solution which is more or less what you suggested:
```
public class ParentDbo extends AuditEntity {
    private final Set<ChildDbo> children = new HashSet<>();

    public void setChildren(Set<ChildDbo> children) {
        this.children.clear();
        if (children != null) {
            this.children.addAll(children);
        }
    }

    public Set<ChildDbo> getChildren() {
        return children;
    }
}
```

```
public interface UpdateMapper<T> {
    void update(T source, @MappingTarget T target);
}
```
```
public interface ExistingCollectionItemsAwareUpdateMapper<T extends AbstractEntity> {

    void update(T source, @MappingTarget T target, @Context UpdateMapper<T> mapper);

    static <T extends AbstractEntity> void updateExistingAware(
            Set<T> source, Set<T> target, @Context UpdateMapper<T> mapper) {
        updateExistingAware(mapper::update, source, target);
    }

    static <T extends AbstractEntity> void updateExistingAware(
            BiConsumer<T, T> updateMapping, Set<T> source, Set<T> target) {
        copyOf(target).stream().filter(t -> !source.contains(t)).forEach(target::remove);
        source.forEach(
                s ->
                        target.stream()
                                .filter(t -> t.equals(s))
                                .findAny()
                                .ifPresentOrElse(t -> updateMapping.accept(s, t), () -> target.add(s)));
    }
}
```
`ExistingCollectionItemsAwareUpdateMapper` can handle hierarchy while `UpdateMapper` is for the "lowest-object".

The child-mapper extends the `ExistingCollectionItemsAwareUpdateMapper` because it's multiple hierarchy otherwise it would be `extends UpdateMapper<ChildDbo>`
```
@Mapper(
        config = BaseConfig.class)
public interface ChildMapper
        extends ExistingCollectionItemsAwareUpdateMapper<ChildDbo> {
}
```
The problem is with the `ExistingCollectionItemsAwareUpdateMapper`. The parent update method looks like that:
```
@Mapper(
        config = BaseConfig.class,
        uses = {ChildMapper.class})
public interface ParentMapper {
    @Mapping(target = "children", ignore = true)
    void update(ParentDbo source, @MappingTarget ParentDbo target, @Context ChildMapper childMapper);
}
```
```
@BeforeMapping
default void updateCollectionItems(ParentDbo source, @MappingTarget ParentDbo target, @Context ChildMapper childMapper) {
        updateExistingAware(source.getChildren(), target.getChildren(), childMapper);
}
```
This is cumbersome because one has to `ignore` first and pass all the child-mappers to the `update`-method as well to the `@BeforeMapping`-method. Also consider this is a simplified example. In case one has multiple levels of hierarchy one has to pass multiple/all `@Context`-mappers on the highest level and pass accordingly. One can lose track easily. If a `@Context`-mapper has been missed, the update-method is automatically created and the custom method is unused (which might result in unwanted behaviour).

When a `update` flag is provided (to pass a update-method) it would be crucial that one does not have to pass the child-mappers via `@Context` but somehow has the mapper accessible in the `BiConsumer`-mapping function (marker/named update mapper?) to call the update-method on that mapper.

I imagine it's not an easy request. I hope this explanation has been helpful. Let me know if I can provide you more information! Thanks for taking a look at it!

**Comment by filiphr** [comment]
>Currently we have this solution which is more or less what you suggested:

The main difference is that you are doing it in `@BeforeMapping` / `@AfterMapping` and what I am proposing is to use a regular custom mapping. This way you should not need to do the `ignore`. MapStruct should then use the regular update methods.

**Comment by lkuendig** [comment]
jep. and for my understanding, that "regular custom mapping" is a referenced method (similar to the `expression`-flag)? how is the update-mapper accessible/passed to that method?

**Comment by lkuendig** [comment]
the problem with the "update"-method only is (instead of the ignore-and-beforeMapping-approach) if there is a bi-directional relationship of the database objects it must be modelled with [controlled modification](https://vladmihalcea.com/jpa-bidirectional-sync-methods). therefore one has to use adder/remove methods instead of setter.

with default `collectionMappingStrategy` mapstruct uses a `clear` and `set` (both not allowed).

with `collectionMappingStrategy = TARGET_IMMUTABLE` and your approach I get the error `Property "xxx" has no write accessor in XXX.`.

therefore a `collectionMappingStrategy` which uses the `ADDER_PREFERRED`-approach without `clear` and `set` around it would be needed

**Comment by filiphr** [comment]
@Nikolas-Charalambidis it seems like what you are asking is quite specific to the use case that you have.

If you use `TARGET_IMMUTABLE` then my approach does not work as the assumption is that you cannot use the getter to change the target collection, that is a feature. If you want to use an adder then I don't see what kind of code MapStruct would need to generate.

**Comment by filiphr** [comment]
@lkuendig, I see what you mean, you basically want `TARGET_IMMUTABLE` with `ADDER_PREFERRED`. Have you tried `TARGET_IMMUTABLE` and no setter?

**Comment by arthurgregorio** [comment]
Maybe I have almost the same issue here

Right now I have a mapping class who does this (full class [here](https://github.com/web-budget/back-end/blob/main/src/main/kotlin/br/com/webbudget/application/mappers/financial/PeriodMovementMapper.kt))

```kotlin
@Mapping(target = "financialPeriod", source = "financialPeriod", qualifiedByName = ["mapFinancialPeriod"])
abstract fun mapToDomain(form: PeriodMovementUpdateForm, @MappingTarget periodMovement: PeriodMovement)
```

the generated method for it is

```java
@Override
public void mapToDomain(PeriodMovementUpdateForm form, PeriodMovement periodMovement) {
    if ( form == null ) {
        return;
    }

    if ( form.getFinancialPeriod() != null ) {
        periodMovement.setFinancialPeriod( mapFinancialPeriod( form.getFinancialPeriod() ) );
    }
    if ( form.getName() != null ) {
        periodMovement.setName( form.getName() );
    }
    if ( form.getDueDate() != null ) {
        periodMovement.setDueDate( form.getDueDate() );
    }
    if ( form.getValue() != null ) {
        periodMovement.setValue( form.getValue() );
    }
    if ( form.getDescription() != null ) {
        periodMovement.setDescription( form.getDescription() );
    }

    if ( periodMovement.getApportionments() != null ) {
        periodMovement.getApportionments().clear();
        List<Apportionment> list = apportionmentFormListToApportionmentList( form.getApportionments() );
        if ( list != null ) {
            periodMovement.getApportionments().addAll( list );
        }
    }
}
```

I have [a test to cover this](https://github.com/web-budget/back-end/blob/main/src/test/kotlin/br/com/webbudget/mappers/financial/PeriodMovementMapperUTest.kt), and is failing

The bad part is this `clear()` in the target class, cause not always the form (input) is coming with all the data we already have, so, for this case we should not touch the current list and add just the new ones. Also another strange thing I've noticed is the fact that mapstruct is checking for nullability of the target field rater than the source field like the other ones. and even if the source property is null (something that could happen) the target will be cleared and the value passed to the service is wrong since them it will remove those entities from the database.

I'm using the default config for mapping collections, my expectation is basically that when the `periodMovement.getApportionments()` has items, it shouldn't be cleared and all the new values should be added.

**Comment by filiphr** [comment]
>The bad part is this clear() in the target class, cause not always the form (input) is coming with all the data we already have, so, for this case we should not touch the current list and add just the new ones

That's by design @arthurgregorio. We are not doing partial updates, we are basically replacing what is there with what comes from the source. The `.clear()` can be done using the `CollectionMappingStrategy#TARGET_IMMUTABLE`, but then `getXX().addAll( ... )` cannot be used.  You can try using that and adding an `adder` method

> Also another strange thing I've noticed is the fact that mapstruct is checking for nullability of the target field rater than the source field like the other ones. and even if the source property is null (something that could happen) the target will be cleared and the value passed to the service is wrong since them it will remove those entities from the database.

This is also by design. This is doing an update and not a patch. You can control what should happen if the source is `null` by using `NullValuePropertyMappingStrategy#IGNORE`.

**Comment by filiphr** [comment]
OK, I see what you mean. The reason why the target check is there is due to the `CollectionMappingStrategy`. Basically, by default if there is no setter, which in your case it seems like there is no `setApportionments`, then MapStruct will use the getter as a setter. An in this case in order to be equivalent to a set, it needs to clear the target collection and add all source elements to it.

That's why there is a null check before. 

>If I not understood it wrong, it should be done before cause them the behavior of clearing the collection is not a problem anymore because it won't be touched if the source is null
>
>Right now if the source is null the collection will be cleared.

It might be an entirely different issue. I don't see how what you are saying is related to the original issue.

In any case, you can add a setter and then I believe it'll work properly and / or use `CollectionMappingStrategy#TARGET_IMMUTABLE`

**Comment by arthurgregorio** [comment]
Got it, I will try this and if it doesn't work, I can open another thread to talk about it.

Thanks for the answers @filiphr !

**Comment by lkuendig** [comment]
> @lkuendig, I see what you mean, you basically want `TARGET_IMMUTABLE` with `ADDER_PREFERRED`. Have you tried `TARGET_IMMUTABLE` and no setter?

you mean something like:
```
@Mapping(target = "children", expression = java(ExistingCollectionItemsAwareUpdateMapper.updateExistingAware(source.getChildren(), target.getChildren(), children))")
void update(ParentDbo source, MappingTarget ParentDbo target, Context ChildrenMapper childrenMapper);
```

with `TARGET_IMMUTABLE` mapstruct (sometimes?) confuses then between `add` and `get` method of that "immutable" object and generates:

```
if ( target.addChildren() != null ) {
        target.addChildren().clear();
        ChildDbo childDbo = ExistingCollectionItemsAwareUpdateMapper.updateExistingAware(source.getChildren(), target.getChildren(), childrenMapper);
        if ( childrenDbo != null ) {
            target.addChildren().putAll( childDbo );
        }
}
```
which does not compile

**Comment by lkuendig** [comment]
is this request still being examined or has it been rejected?

**Comment by filiphr** [comment]
It's been a while @lkuendig. If you are referring to https://github.com/mapstruct/mapstruct/issues/3742#issuecomment-2545586343, then perhaps we need to create another issue for this particular case. The `addChildren` in the example there is actually a fluent getter. So ideally we have some MCVE that we can run and try it out

**Comment by lkuendig** [comment]
ok, so the proposed way of mapstruct to update existing items will be by `expression = java(...)` and a certain strategy?

then I will create another issue for that with a MCVE 👍
